File: src/main/java/com/example/message/service/UserService.java
package com.example.message.service;

import com.example.message.domain.FirebaseAuthToken;
import com.example.message.domain.User;
import com.example.message.domain.dto.UserDto;
import com.example.message.mapper.UserMapper;
import com.example.message.repository.FirebaseAuthTokenRepository;
import com.example.message.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class UserService {

    @Autowired
    public UserService(UserRepository repository, UserMapper mapper, FirebaseAuthTokenRepository firebaseRepository) {
        this.repository = repository;
        this.mapper = mapper;
        this.firebaseRepository = firebaseRepository;
    }

    public boolean createUser(UserDto userDto) {
        if (repository.findByMail(userDto.getMail()) != null) {
            return false;
        }
        User user = mapper.mapToDomain(userDto);
        repository.save(user);
        return true;
    }

    public UserDto loginUser(UserDto userDto) {
        User user = repository.findByMail(userDto.getMail());
        if (user != null && user.getPassword().equals(userDto.getPassword())) {
            return mapper.mapToDto(user);
        }
        return null;
    }

    public UserDto getUserByMail(String mail) {
        User user = repository.findByMail(mail);
        if (user != null) {
            return mapper.mapToDto(user);
        }
        return null;
    }


    public void addFirebaseAuthToken(long userId, String authToken) {
        Optional<User> optionalUser = repository.findById(userId);
        if (optionalUser.isPresent()) {
            User user = optionalUser.get();
            boolean contains = user.getFirebaseAuthTokens().stream().anyMatch(a -> a.getFirebaseAuthToken().equals(authToken));
            if (contains) {
                return;
            }
            List<FirebaseAuthToken> tokens = firebaseRepository.findAll().stream().filter(a -> a.getFirebaseAuthToken().equals(authToken)).collect(Collectors.toList());
            for (FirebaseAuthToken token: tokens) {
                firebaseRepository.delete(token);
            }
            FirebaseAuthToken token = firebaseRepository.save(new FirebaseAuthToken(authToken));
            user.getFirebaseAuthTokens().add(token);
            repository.save(user);
        }
    }

    public List<UserDto> getAllUsers() {
        return mapper.mapToDtoList(repository.findAll());
    }

    private final UserRepository repository;
    private final UserMapper mapper;
    private final FirebaseAuthTokenRepository firebaseRepository;
}

File: src/main/java/com/example/message/service/UserService.kt
package com.example.message.service

import com.example.message.domain.FirebaseAuthToken
import com.example.message.domain.dto.UserDto
import com.example.message.mapper.UserMapper
import com.example.message.repository.FirebaseAuthTokenRepository
import com.example.message.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import java.util.stream.Collectors

@Service
class UserService(
        private val repository: UserRepository,
        private val mapper: UserMapper,
        private val firebaseRepository: FirebaseAuthTokenRepository
) {

    fun createUser(userDto: UserDto): Boolean {
        repository.findByMail(userDto.mail)?.let {
            return false
        }
        val user = mapper.mapToDomain(userDto)
        repository.save(user)
        return true
    }

    fun loginUser(userDto: UserDto): UserDto? {
        repository.findByMail(userDto.mail)?.let {
            if (it.password == userDto.password) {
                return mapper.mapToDto(it)
            }
        }
        return null
    }

    fun getUserByMail(mail: String): UserDto? {
        repository.findByMail(mail)?.let {
            mapper.mapToDto(it)
        }
        return null
    }

    fun addFirebaseAuthToken(userId: Long, authToken: String) {
        repository.findByIdOrNull(userId)?.let {user ->
            val contains = user.firebaseAuthTokens.any { a -> a.firebaseAuthToken == authToken }
            if (contains) {
                return
            }
            val tokens = firebaseRepository.findAll().stream().filter { a: FirebaseAuthToken -> a.firebaseAuthToken == authToken }.collect(Collectors.toList())
            for (token in tokens) {
                firebaseRepository.delete(token)
            }
            val token = firebaseRepository.save(FirebaseAuthToken(-1, authToken))
            user.firebaseAuthTokens.add(token)
            repository.save(user)
        }
    }

    val allUsers: List<UserDto>
        get() = mapper.mapToDtoList(repository.findAll())

}